src/streaming/thread.pool.cpp

```diff
@@ -0,0 +1,96 @@
+#include "thread.pool.hh"
+
+zarr::ThreadPool::ThreadPool(unsigned int n_threads,
+                             std::function<void(const std::string&)> err)
+  : error_handler_{ err }
```

Comment:
`std::function` can be costly in terms of performance. Moreover, it's common practice to use inline lambdas for scenarios like error callbacks. I recommend shifting to function objects to address these situations more efficiently.

```cpp
: error_handler_ { std::move(err) }
```

Reply:
Something like this?

```c++
class ErrorCallerbacker {
public:
    ErrorCallbacker(std::unique_ptr<ZarrStream_s> stream) : stream_(stream) {}
    int operator()(const std::string& s) const { stream_->set_error(s); }
private:
    int stream_;
};
```

Reply:
Ignore the last part of my comment. My primary point was moving the object instead of copying it.